cross-version-test/cross_version_test/cli.py

```diff
@@ -0,0 +1,276 @@
+    click.echo("Done!")
+
+
+def list_jobs() -> t.List[str]:
+    docker_compose = DockerCompose(DOCKER_COMPOSE_FILE)
+    return docker_compose.config("--services", capture_output=True, check=True).stdout.splitlines()
+
+
+@cli.command(name="list")
+def list_command():
+    click.echo("\n".join(list_jobs()))
+
+
+@cli.command(context_settings=dict(ignore_unknown_options=True))
+@click.option(
+    "--pattern",
+    "-p",
+    required=False,
+    help="If specified, only run jobs that matches the given pattern (regular expression). "
+    "For example, 'xgboost_.+_autologging' matches xgboost autologging jobs.",
+)
+@click.option(
+    "--no-cache", is_flag=True, default=False, help="Do not use cache when building images"
+)
+def build(pattern: t.Optional[str], no_cache: bool):
+    jobs = list_jobs()
+
+    no_cache_opt = ["--no-cache"] if no_cache else []
+
+    # Build base image
+    reporter = Reporter()
+    reporter.write_separator("Building base image")
+    this_file_dir = Path(__file__).parent
+    base_dockerfile = this_file_dir.joinpath("base_image", "Dockerfile")
+    run_cmd(
+        [
+            "docker",
+            "build",
+            "-f",
+            base_dockerfile,
+            *no_cache_opt,
+            "-t",
+            c.DOCKER_BASE_IMAGE,
+            this_file_dir,
+        ],
+        check=True,
+    )
+
+    if pattern:
+        jobs = [j for j in jobs if re.compile(pattern).search(j)]
+
+    docker_compose = DockerCompose(DOCKER_COMPOSE_FILE)
+    for job in jobs:
+        reporter.write_separator(f"Building {job}")
+        prc = docker_compose.build(*no_cache_opt, job, check=False)
+        reporter.add_result(Result(job, prc.success))
+
+    reporter.write_separator("Results")
+    reporter.report_results()
+    failed = not reporter.all_success()
+    exit(failed)
+
+
+@cli.command(context_settings=dict(ignore_unknown_options=True))
+@click.option(
+    "--pattern",
+    "-p",
+    required=False,
+    help="If specified, only run jobs that matches the given pattern (regular expression). "
+    "For example, 'xgboost_.+_autologging' matches xgboost autologging jobs.",
+)
+@click.argument("cmd", nargs=-1)
+def run(pattern: t.Optional[str], cmd: t.Tuple[str, ...]):
+    jobs = list_jobs()
+
+    if pattern:
+        jobs = [j for j in jobs if re.compile(pattern).search(j)]
+
+    docker_compose = DockerCompose(DOCKER_COMPOSE_FILE)
+    reporter = Reporter()
+    for job in jobs:
+        reporter.write_separator(f"Running {job}")
+        prc = docker_compose.run("--rm", job, *cmd, check=False)
+        reporter.add_result(Result(job, prc.success))
+
+    reporter.write_separator("Results")
+    reporter.report_results()
+    failed = not reporter.all_success()
+    exit(failed)
+
+
+@cli.command(context_settings=dict(ignore_unknown_options=True))
+@click.option(
+    "--rmi",
+    is_flag=True,
+    default=False,
+    help="If True, remove images created by cross-version test",
+)
+def down(rmi):
```

Comment:
What do you think about making a 'soft delete' approach similar to pruning activities (only remove volumes, containers, images if they're older than "x hours" so that if someone is troubleshooting and forgets to remove older images, a call to clean the environment can be made during initialization to check to see if there are older images and get rid of them?) e.g., https://docs.docker.com/config/pruning/ ?

Reply:
Certainly! This was just an idea for future work (if it's needed).